src/treelearner/serial_tree_learner.cpp

```diff
@@ -343,9 +348,20 @@ void SerialTreeLearner::ConstructHistograms(
     const std::vector<int8_t>& is_feature_used, bool use_subtract) {
   Common::FunctionTimer fun_timer("SerialTreeLearner::ConstructHistograms",
                                   global_timer);
+  int offset = 1;
+  if (!share_state_->is_colwise && !share_state_->multi_val_bin->IsSparse() &&
+    train_data_->FeatureBinMapper(0)->GetMostFreqBin() > 0 &&
+    train_data_->IsMultiGroup(train_data_->Feature2Group(0))) {
+    offset = 0;
+  }
+  if (share_state_->is_colwise &&
+    train_data_->IsDenseMultiGroup(train_data_->Feature2Group(0)) &&
+    train_data_->FeatureBinMapper(0)->GetMostFreqBin() > 0) {
+    offset = 0;
+  }
   // construct smaller leaf
   hist_t* ptr_smaller_leaf_hist_data =
-      smaller_leaf_histogram_array_[0].RawData() - kHistOffset;
+      smaller_leaf_histogram_array_[0].RawData() - kHistOffset * offset;
```

Comment:
I think the histogram data is always offset by 1, refer to line 1254 in feature_histogram.hpp.
```
pool_[i][j].Init(data_[i].data() + offsets[j] * 2, &feature_metas_[j]);
```